src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,282 @@
+#include "src/core/lib/slice/slice_buffer.h"
+
+namespace grpc_core {
+
+// Wrapper around event engine endpoint that provides a promise like API.
+class PromiseEndpoint {
+ public:
+  PromiseEndpoint(
+      std::unique_ptr<grpc_event_engine::experimental::EventEngine::Endpoint>
+          endpoint,
+      SliceBuffer already_received);
+  ~PromiseEndpoint();
+
+  // Returns a promise that resolves to a `absl::Status` indicating the result
+  // of the write operation.
+  //
+  // Concurrent writes are not supported, which means callers should not call
+  // `Write()` before the previous write finishes. Doing that results in
+  // undefined behavior.
+  auto Write(SliceBuffer data) {
+    {
+      MutexLock lock(&write_mutex_);
+      // Previous write result has not been polled.
+      GPR_ASSERT(!write_result_.has_value());
+      // TODO(ladynana): Replace this with `SliceBufferCast<>` when it is
+      // available.
+      grpc_slice_buffer_swap(write_buffer_.c_slice_buffer(),
+                             data.c_slice_buffer());
+    }
+    // If `Write()` returns true immediately, the callback will not be called.
+    // We still need to call our callback to pick up the result.
+    if (endpoint_->Write(std::bind(&PromiseEndpoint::WriteCallback, this,
+                                   std::placeholders::_1),
+                         &write_buffer_,
+                         nullptr /* uses default arguments */)) {
+      WriteCallback(absl::OkStatus());
+    }
+    return [this]() -> Poll<absl::Status> {
+      MutexLock lock(&write_mutex_);
+      if (!write_result_.has_value()) {
+        write_waker_ = Activity::current()->MakeNonOwningWaker();
+        return Pending();
+      } else {
+        const auto ret = *write_result_;
+        write_result_.reset();
+        return ret;
+      }
+    };
+  }
+
+  // Returns a promise that resolves to `SliceBuffer` with
+  // `num_bytes` bytes.
+  //
+  // Concurrent reads are not supported, which means callers should not call
+  // `Read()` before the previous read finishes. Doing that results in
+  // undefined behavior.
+  auto Read(size_t num_bytes) {
+    ReleasableMutexLock lock(&read_mutex_);
+    // Previous read result has not been polled.
+    GPR_ASSERT(!read_result_.has_value());
+    // Should not have pending reads.
+    GPR_ASSERT(pending_read_buffer_.Count() == 0u);
+    if (read_buffer_.Length() < num_bytes) {
+      lock.Release();
+      // If `Read()` returns true immediately, the callback will not be
+      // called. We still need to call our callback to pick up the result and
+      // maybe do further reads.
+      if (endpoint_->Read(std::bind(&PromiseEndpoint::ReadCallback, this,
+                                    std::placeholders::_1, num_bytes,
+                                    absl::nullopt /* uses default arguments */),
+                          &pending_read_buffer_,
+                          nullptr /* uses default arguments */)) {
```

Comment:
Probably can be improved by providing the EventEngine::Endpoint::ReadArgs optional argument with num_bytes as hint. Some EE implementations (e.g Posix) will ensure that the Read doesn't complete internally without reading the  minimum required number of bytes.

Reply:
Done, thanks for the suggestion.